commit 39c1abd0d4e6ea54560d4de38d0d058e3bebefc7 (refs/changes/37/1237/3)
Author: David Pacheco <dap@joyent.com>
Date:   2017-01-11T13:22:46-08:00 (2 years, 9 months ago)
    
    MANTA-3080 muskie picker only picks from 200 nodes
    Reviewed by: Bryan Cantrill <bryan@joyent.com>
    Approved by: Bryan Cantrill <bryan@joyent.com>
    Approved by: Jordan Hendricks <jordan.hendricks@joyent.com>

diff --git a/lib/picker.js b/lib/picker.js
index 5f72e72..6c0bc35 100644
--- a/lib/picker.js
+++ b/lib/picker.js
@@ -82,6 +82,7 @@ var fetch = function fetch_moray(opts, cb) {
     cb = once(cb);
 
     var count = 0;
+    var recs = 0;
     var f = sprintf('(&(percentUsed<=%d)(timestamp>=%d)%s)',
                     opts.utilization || 90,
                     Date.now() - opts.lag,
@@ -103,10 +104,17 @@ var fetch = function fetch_moray(opts, cb) {
         values.push(data.value);
         count = data._count;
         marker = data._id;
+        recs++;
     });
 
     req.once('end', function () {
-        if (values.length < count) {
+        /*
+         * We only fetch "limit" records, but there may be many more storage
+         * nodes than that.  If we saw fewer records than the number that Moray
+         * reported matched our query, that means there are more to fetch, so
+         * we take another lap.
+         */
+        if (recs < count) {
             var next = {
                 lag: opts.lag,
                 limit: opts.limit,
